### PR TITLE
Release public weights/model checkpoint for multi-task PoyoPlus

### PR DESCRIPTION
**Pull request by ijight**
- Fixed a bug in running speed computation where .values() was incorrectly set to .timestamps().
- Updated the implementation of multitask loss weighting.
- Added a loss term for natural movies so that a prediction is considered correct if it matches the target anywhere within the context window (e.g., across the 30-frame / 1-second span).

[Model Checkpoint](https://drive.google.com/file/d/1zKzR9YU1dDfqe9ygcBLfdj8Masjk_I-l/view?usp=sharing )

| Metric                                                   | Value                     |
|----------------------------------------------------------|---------------------------|
| average_test_metric                                      | 0.33522698283195496       |
| test/average_drifting_gratings_orientation_metric        | 0.4719316363334656        |
| test/average_drifting_gratings_temporal_frequency_metric | 0.45032551884651184       |
| test/average_locally_sparse_noise_frame_metric           | 2.451429509164882e-06     |
| test/average_natural_movie_one_frame_metric              | 0.4296070635318756        |
| test/average_natural_movie_three_frame_metric            | 0.15975013375282288       |
| test/average_natural_movie_two_frame_metric              | 0.36854615807533264       |
| test/average_natural_scenes_metric                       | 0.09675408899784088       |
| test/average_pupil_location_metric                       | 0.2790534496307373        |
| test/average_running_speed_metric                        | 0.4021194279193878        |
| test/average_static_gratings_orientation_metric          | 0.33467045426368713       |
| test/average_static_gratings_phase_metric                | 0.26640233397483826       |
| test/average_static_gratings_spatial_frequency_metric    | 0.4454978108406067        |

These vary from those reported in the paper. Running speed drops. Most improve. Natural scenes doubles.

## Summary by CodeRabbit

* **New Features**
  * Introduced CalciumPOYOPlus model with public pretrained weights
  * Added per-task average metric logging during validation
  * Automatic multi-task dataset configuration generation from tabular input
  * New tolerance-based accuracy metric for natural movie predictions

* **Improvements**
  * Enhanced checkpoint management with regular evaluation-based saves
  * Training and validation now respect worker configuration settings

* **Documentation**
  * Updated guides with new dataset generation workflow and pretrained weights integration

### CODE REVIEW

File: run_capoyo.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+#SBATCH --job-name=capoyo-full
+#SBATCH --nodes=1
+#SBATCH --gpus=4
+#SBATCH --mem-per-gpu=128G
+#SBATCH --cpus-per-gpu=16
+#SBATCH --output=logs/capoyo-full-%j.out
+#SBATCH --error=logs/capoyo-full-%j.err
+source /vast/projects/dyer1/lab/user/ian/envs/torch_brain_env/bin/activate
```
Comment (mazabou) [inline run_capoyo.sh]: hardcoded path

File: run_capoyo.sh
Comment (mazabou) [inline run_capoyo.sh:1]: rename capoyo > poyo_plus

File: torch_brain/utils/stitcher.py [hunk 2]
```diff
@@ -82,28 +101,389 @@ def stitch(
...
+                    #     if torch_brain.get_modality_by_id(task_name).type == DataType.CONTINUOUS:
+                    #         self.metrics[recording_id][task_name][metric_name].to(pl_module.device).update(
+                    #             torch.empty(0, dim, device=pl_module.device),
+                    #             torch.empty(0, dim, device=pl_module.device),
+                    #         )
+                    #     else:  # Classifier
+                    #         self.metrics[recording_id][task_name][metric_name].to(pl_module.device).update(
+                    #             torch.empty(1, dim, 0, device=pl_module.device),
+                    #             torch.empty(1, 0, device=pl_module.device),
+                    #         )
```
Comment (mazabou) [inline torch_brain/utils/stitcher.py]: this was added by mistake?

File: torch_brain/metrics/frame_index_window_accuracy.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+"""Metrics for ordinal frame indices (e.g. natural movie readouts)."""
+
+import torch
+from torch import Tensor
+from torchmetrics import Metric
+
+
+class FrameIndexWindowAccuracy(Metric):
```
Comment (mazabou) [inline torch_brain/metrics/frame_index_window_accuracy.py]: maybe call it WithinDeltaAccuracy?

File: torch_brain/metrics/frame_index_window_accuracy.py
Comment (vinamarora8) [inline torch_brain/metrics/frame_index_window_accuracy.py:1]: Do you use the contents of this module anywhere?

File: run_capoyo.sh
Comment (vinamarora8) [inline run_capoyo.sh:1]: This script is at project root level Ian.. I would suggest moving instructions for running the example in the example's README

File: torch_brain/utils/stitcher.py
Comment (vinamarora8) [inline torch_brain/utils/stitcher.py:1]: This file has merge conflicts with current `main`

File: examples/poyo_plus/scripts/EDA.ipynb
Comment (vinamarora8) [inline examples/poyo_plus/scripts/EDA.ipynb:1]: What does this notebook do?

File: examples/poyo_plus/configs/dataset/capoyo.yaml
Comment (vinamarora8) [inline examples/poyo_plus/configs/dataset/capoyo.yaml:1]: Maybe we can replace this yaml file with instructions for how to run `generate_config.py`? (I'm assuming it creates this yaml file)

**Review by vinamarora8 — CHANGES_REQUESTED** [review]
Thanks @ijight. Can you fix the merge conflict. Also left some comments

File: run_capoyo.sh
Comment (ijight) [reply run_capoyo.sh:1]: I will just remove it i think

File: torch_brain/utils/stitcher.py
Comment (ijight) [reply torch_brain/utils/stitcher.py:1]: Yah this has the old fix for the broken val stitcher

File: examples/poyo_plus/scripts/EDA.ipynb
Comment (ijight) [reply examples/poyo_plus/scripts/EDA.ipynb:1]: I think I will remove it

File: examples/poyo_plus/configs/dataset/capoyo.yaml
Comment (ijight) [reply examples/poyo_plus/configs/dataset/capoyo.yaml:1]: Yah that's probably better

**Comment by ijight** [comment]
Heads up the task registry has changed so I don't think this works out of the box?

File: examples/poyo_plus/metrics.py
Comment (vinamarora8) [inline examples/poyo_plus/metrics.py:1]: I would say introducing a whole new module would deserve its own PR and longer discussion. So, maybe just move this metric within the `examples/poyo_plus` directory for now?

If it proves to be more generally applicable over time, we can create a root level module for it later.

File: torch_brain/registry.py [hunk 4]
```diff
@@ -108,6 +108,19 @@ def get_modality_by_id(modality_id: int) -> ModalitySpec:
...
+    Args:
+        modality_name: The name of the modality to retrieve
+
+    Returns:
+        ModalitySpec: The modality specification
+    """
+    if modality_name not in MODALITY_REGISTRY:
+        raise KeyError(f"No modality found with name {modality_name}")
+    return MODALITY_REGISTRY[modality_name]
+
```
Comment (vinamarora8) [inline torch_brain/registry.py]: Why not just use `MODALITY_REGISTRY[modality_name]` where you need this fn?

File: examples/poyo_plus/README.md [hunk 5]
```diff
@@ -6,32 +6,82 @@ Official codebase for POYO+ from [ICLR 2025].
 ---
 
 POYO+ is a multi-task version of POYO.
-This is an example training script for the model in [Azabou and Pan et al. 2025](https://proceedings.iclr.cc/paper_files/paper/2025/file/953390c834451505703c9da45de634d8-Paper-Conference.pdf), corresonding to the module `torch_brain.models.CaPOYO`.
+This is an example training script for the model in [Azabou and Pan et al. 2025](https://proceedings.iclr.cc/paper_files/paper/2025/file/953390c834451505703c9da45de634d8-Paper-Conference.pdf), corresonding to the module `torch_brain.models.CalciumPOYOPlus`.
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo_plus/README.md]: @ijight valid typo

File: torch_brain/registry.py (hunk 4 above)
Comment (ijight) [reply torch_brain/registry.py]: I just like the symmetry but I suppose there's no reason for it.

File: examples/poyo_plus/README.md
Comment (vinamarora8) [inline examples/poyo_plus/README.md:1]: README looking very nice!

**Review by vinamarora8 — COMMENTED** [review]
Looks good on first pass now! I will do another read through tomorrow. For now:

Can you please add 3 things to the CHANGELOG:
1. [Added] Added public weights for CalciumPoyoPlus
2. [Changed] Renamed `torch_brain.models.CaPOYO` to `torch_brain.models.CalciumPoyoPlus`
3. [Added] Added per-task average metric logging in `MultitaskDecodingStitchEvaluator`

Also, please fix the CI issues!

File: examples/poyo_plus/train.py [hunk 6]
```diff
@@ -409,7 +413,7 @@ def main(cfg: DictConfig):
     )
 
     # Train
-    trainer.fit(wrapper, data_module, ckpt_path=cfg.ckpt_path)
+    trainer.fit(wrapper, data_module, ckpt_path=cfg.ckpt_path, weights_only=False)
```
[the remark this answers is not in the document]
Comment (vinamarora8) [reply examples/poyo_plus/train.py:422]: @ijight I guess would be better to change to `lightning>=2.6.0` in the example's README.md

**Comment by vinamarora8** [comment]
Can you also add the results table that is in this PR's description to the example's README please?
Would be ideal to also include small explanation for why these numbers differ from the paper.

File: examples/poyo_plus/README.md [hunk 7]
```diff
@@ -6,16 +6,19 @@ Official codebase for POYO+ from [ICLR 2025].
...
 
 **Installing necessary packages**
 ```bash
-pip install torch_brain lightning wandb brainsets
+pip install torch_brain "lightning>=2.6.0" wandb brainsets
 ```
 
+> Note: `lightning>=2.6.0` is required because `train.py` passes the
+> `weights_only=False` argument to `Trainer.fit`, which was added in 2.6.0.
+
```
Comment (vinamarora8) [inline examples/poyo_plus/README.md:21]: You probably meant to do
```
> [!NOTE]
> `lightning>=2.6.0` is required because `train.py` passes the
> `weights_only=False` argument to `Trainer.fit`, which was added in 2.6.0.
```

File: examples/poyo_plus/README.md [hunk 8]
```diff
@@ -64,6 +67,31 @@ The `unit_emb` and `session_emb` `InfiniteVocabEmbedding` layers will be
...
+| natural_movie_three_frame            | mean(Accuracy, WithinDeltaAcc)    | 0.1598 |
+| pupil_location                       | MeanSquaredError (z-normalized)   | 0.2791 |
+| running_speed                        | MeanSquaredError (z-normalized)   | 0.4021 |
+
+> Note: numbers differ from those reported in the paper. In particular, our
+> `running_speed` prediction target was actually the monotonically-increasing
+> timestamps rather than the m/s running-speed value, so that row is not
+> directly comparable. The remaining differences are attributable to small
+> hyperparameter/implementation deltas and stochasticity in training.
+
```
Comment (vinamarora8) [inline examples/poyo_plus/README.md:94]: Same here

**Review by vinamarora8 — COMMENTED** [review]
Thanks for the edits! Looking good. Left 2 minor comments regarding README.

**Review by vinamarora8 — APPROVED** [review]
Approving and merging quickly! The minor issue in README can be resolved with another small PR

**Comment by ijight** [comment]
Should also add the calcium poyo paper to the main readme

**Comment by vinamarora8** [comment]
@ijight I made the readme edits in #202. Could you please review them?
